File: backend/src/sidra/api/models/roadmap_row.py
from __future__ import annotations

import uuid
from datetime import date

from pydantic import BaseModel


class RoadmapRow(BaseModel):
    """A dated projection for one track.

    ``projected_finish`` already carries the current debt: every day owed slides the date by a day,
    which is the whole reason the roadmap is worth looking at.
    """

    track_id: uuid.UUID
    name_en: str
    name_he: str
    work_ref_title: str | None
    """The work the track is standing in. A track named "Gemara" that holds one masechta projects
    that masechta, and a row saying only "Gemara" overclaims by a factor of thirty-five."""

    corpus_en: str | None
    """The whole body that work belongs to, when it is larger than the track."""

    corpus_total: int | None
    corpus_years: float | None
    """How long the whole body would take at this track's rate. The honest second scale."""

    total: int
    actual_ordinal: int
    units_remaining: int
    rate_per_day: float
    debt: int
    projected_finish: date | None
    """None on a chavrusa track, which has no rate to project from."""

    yearly_cycle_rate: float
    """How many units a day a full cycle in a year would take. The Pace Explorer's number."""

File: frontend/src/types/RoadmapRow.ts
export interface RoadmapRow {
  readonly track_id: string;
  readonly name_en: string;
  readonly name_he: string;
  /** The work the track is standing in. A row saying only "Gemara" overclaims by a factor of 35. */
  readonly work_ref_title: string | null;
  /** The whole body that work belongs to, when it is larger than the track. */
  readonly corpus_en: string | null;
  readonly corpus_total: number | null;
  /** How long the whole body would take at this track's rate. The honest second scale. */
  readonly corpus_years: number | null;
  readonly total: number;
  readonly actual_ordinal: number;
  readonly units_remaining: number;
  readonly rate_per_day: number;
  readonly debt: number;
  /** Null on a chavrusa track, which has no rate to project from. */
  readonly projected_finish: string | null;
  /** How many units a day a full cycle in a year would take. The Pace Explorer's number. */
  readonly yearly_cycle_rate: number;
}
